fw/cache.c

```diff
@@ -328,21 +328,51 @@ tfw_cache_key_node(unsigned long key)
 	return key % num_online_nodes();
 }
 
+/**
+ * Release memory dynamically allocated
+ * for every node cpus
+ */
+static void
+tfw_release_node_cpus(void)
+{
+	int node;
+
+	for_each_online_node(node) {
+		if(c_nodes[node].nr_cpus > 0)
+			kfree(c_nodes[node].cpu);
+	}
+}
+
 /**
  * Just choose any CPU for each node to use queue_work_on() for
  * nodes scheduling. Reserve 0th CPU for other tasks.
  */
-static void
+static int
 tfw_init_node_cpus(void)
 {
-	int cpu, node;
+	int nr_cpus, cpu, node;
+
+	nr_cpus = num_online_cpus();
 
 	for_each_online_cpu(cpu) {
+
+
 		node = cpu_to_node(cpu);
+
+		pr_err("node.nr_cpus %u",c_nodes[node].nr_cpus);
+
+		c_nodes[node].cpu = kmalloc(nr_cpus * sizeof(int), GFP_KERNEL);
```

Comment:
This is very wrong! Say you have 64 CPUs in 2 nodes, each node (CPU processor package) has 32 CPU cores (maybe with hyperthreads). In this loop you make 64 allocations and constantly rewrite one of the two `cpu` pointers, leaking the memory.

The initial loop is essentially maps global cpu ID (as you can observe as `processor` in `/proc/cpuinfo`) to a relative CPU id within a node. To make the code simple I'd propose you to rewrite it in 2 loops: one for each node to allocate `cpu` and the second one for each online CPU to write the mappings.

Also I suggest to add an assertion like `BUG_ON(c_nodes[node].nr_cpus > nr_cpus_node(node))`

Reply:
No. We allocate CPU array for every node, than fill it by CPU IDs. So  in case of 2 Node * 32 CPU we'll perform 2 allocations.
But indeed, in current code arrays are oversized. I'll change it size to `nr_cpus_node`, as well as will add `BUG_ON`.

Reply:
@krizhanovsky is right here! Just print node is always 0 here. c_nodes[node].cpu = kmalloc(nr_cpus * sizeof(int), GFP_KERNEL); is always rewrite pointer!